checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py

```diff
@@ -137,30 +138,28 @@ def to_fragments(self) -> fragments.Fragments:
     return result
 
 
-def _hashable_slices(slices: Index) -> tuple[HashableSlice, ...]:
-  return tuple([(s.start, s.stop, s.step) for s in slices])
-
-
-def _create_replica_counts_builder(
-    arr: jax.Array,
-) -> Callable[[jax.Shard], int]:
-  """Produces a mapping from addressable shards to global replication count."""
-
-  counts = collections.defaultdict(int)
-  for index in arr.sharding.devices_indices_map(arr.shape).values():
-    counts[_hashable_slices(index)] += 1
-
-  return lambda shard: counts[_hashable_slices(shard.index)]
+def _sharding_num_replicas(
+    sharding: jax.sharding.Sharding,
+    global_shape: Shape
+) -> int:
+  # The implementation below mirrors jax.sharding.Sharding#devices_indices_map,
+  # but takes care _not_ to materialize data for every device in the mesh, so
+  # that its runtime does not grow with the size of the mesh.
+  sharding.shard_shape(global_shape)  # raises a good error message
+  hlo_sharding = sharding._to_xla_hlo_sharding(len(global_shape))
+  if op_shardings.is_op_sharding_replicated(hlo_sharding):
+    return len(sharding._device_assignment)
+  _, num_replicas = op_shardings.get_num_ways_dim_sharded(hlo_sharding)
```

Comment:
Please don't use JAX internal APIs.

You can use pspec and mesh to calculate how many ways you are replicated across (and error for other shardings).

Also don't replicate get_num_ways_dim_sharded logic here since that also involves using JAX internal APIs.

Reply:
> You can use pspec and mesh to calculate how many ways you are replicated across (and error for other shardings).

Happy to go this way, but to cover all kinds of shardings we'd have to pattern-match on the subclasses of `jax.sharding.Sharding` and re-implement the logic based on their various representations. `NamedSharding` seems straightforward enough; I don't have a great handle on how commonly used `PmapSharding`, `PositionalSharding`, and `GSPMDSharding` are these days. It also seems for the latter two it'd be impossible to access the necessary information through the public API?

Reply:
You can error out for any other sharding than NamedSharding. We are going to deprecate all of those soon. Only NamedSharding will remain.